polaris-react/src/components/Scrollable/Scrollable.tsx

```diff
@@ -36,218 +32,86 @@ export interface ScrollableProps extends React.HTMLProps<HTMLDivElement> {
-      </ScrollableContext.Provider>
-    );
-  }
-
-  private setScrollArea = (scrollArea: HTMLElement | null) => {
-    this.scrollArea = scrollArea;
-  };
-
-  private handleScroll = () => {
-    const {scrollArea} = this;
-    const {scrollPosition} = this.state;
-    const {shadow, onScrolledToBottom} = this.props;
-    if (scrollArea == null) {
+    if (!currentScrollArea) {
       return;
     }
-    const {scrollTop, clientHeight, scrollHeight} = scrollArea;
-    const shouldBottomShadow = Boolean(
-      shadow && !(scrollTop + clientHeight >= scrollHeight),
-    );
-    const shouldTopShadow = Boolean(
-      shadow && scrollTop > 0 && scrollPosition > 0,
-    );
 
-    const canScroll = scrollHeight > clientHeight;
-    const hasScrolledToBottom =
-      scrollHeight - scrollTop <= clientHeight + LOW_RES_BUFFER;
+    const handleScroll = () => {
+      const {scrollTop, clientHeight, scrollHeight} = currentScrollArea;
 
-    if (canScroll && hasScrolledToBottom && onScrolledToBottom) {
-      onScrolledToBottom();
-    }
+      setBottomShadow(
+        Boolean(shadow && !(scrollTop + clientHeight >= scrollHeight)),
+      );
+      setTopShadow(Boolean(shadow && scrollTop > 0));
+    };
 
-    this.setState({
-      topShadow: shouldTopShadow,
-      bottomShadow: shouldBottomShadow,
-      scrollPosition: scrollTop,
-      canScroll,
-    });
-  };
+    const handleResize = debounce(handleScroll, 50, {trailing: true});
 
-  private scrollHint = () => {
-    const {scrollArea} = this;
-    if (scrollArea == null) {
-      return;
-    }
-    const {clientHeight, scrollHeight} = scrollArea;
-    if (
-      PREFERS_REDUCED_MOTION ||
-      this.state.scrollPosition > 0 ||
-      scrollHeight <= clientHeight
-    ) {
-      return;
-    }
+    stickyManager.current?.setContainer(currentScrollArea);
+    currentScrollArea.addEventListener('scroll', handleScroll);
+    window.addEventListener('resize', handleResize);
 
-    const scrollDistance = scrollHeight - clientHeight;
-    this.toggleLock();
-    this.setState(
-      {
-        scrollPosition:
-          scrollDistance > MAX_SCROLL_DISTANCE
-            ? MAX_SCROLL_DISTANCE
-            : scrollDistance,
-      },
-      () => {
-        window.requestAnimationFrame(this.scrollStep);
-      },
-    );
-  };
+    handleScroll();
 
-  private scrollStep = () => {
-    this.setState(
-      ({scrollPosition}) => {
-        const delta = scrollPosition * DELTA_PERCENTAGE;
-        return {
-          scrollPosition: delta < DELTA_THRESHOLD ? 0 : scrollPosition - delta,
-        };
-      },
-      () => {
-        if (this.state.scrollPosition > 0) {
-          window.requestAnimationFrame(this.scrollStep);
-        } else {
-          this.toggleLock(false);
-        }
-      },
-    );
-  };
+    return () => {
+      currentScrollArea.removeEventListener('scroll', handleScroll);
+      window.removeEventListener('resize', handleResize);
+    };
+  }, [shadow]);
```

Comment:
My only minor concern with the code in this PR is that the `shadow` work is encapsulated within a useEffect when it seems like it shouldn't have to be. It's probably fine so not a blocker, it's kind of nice this way anyway but might be causing an unnecessary render, sometimes.